### PR TITLE
adapt CFRClassifier to dataset and `tech.ml`

### PR DESCRIPTION
**Pull request by behrica**

### CODE REVIEW

File: deps.edn [hunk 1]
```diff
@@ -2,7 +2,13 @@
  :deps    {org.clojure/clojure               {:mvn/version "1.10.1"}
            aysylu/loom                       {:mvn/version "1.0.2"}
            camel-snake-kebab                 {:mvn/version "0.4.2"}
-           edu.stanford.nlp/stanford-corenlp {:mvn/version "4.2.0"}}
+           edu.stanford.nlp/stanford-corenlp {:mvn/version "4.2.0"}
+           techascent/tech.ml {:mvn/version "5.01"}
```
Comment (behrica) [inline deps.edn:6]: these extra deps are needed

File: src/dk/simongray/datalinguist/ml/crf.clj [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  (-> dataset
+      (ds/add-column
+       (ds/new-column :ner
+                      (map
+                       #(.classifyToCharacterOffsets crf %)
+                       (get dataset text-col))))))
+
+
+
+(defn train [feature-ds target-ds options]
```
Comment (behrica) [inline src/dk/simongray/datalinguist/ml/crf.clj:96]: This train / test follow the requited signature of `tech.ml`

File: src/dk/simongray/datalinguist/ml/crf.clj [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+   (ds/concat feature-ds target-ds)
+   (first (ds/column-names feature-ds))
+   (first (ds/column-names target-ds))
+   (merge standard-props (:cfr-options options))))
+
+
+(defn predict [feature-ds thawed-model model]
+  (predict-crf feature-ds (first  (ds/column-names feature-ds)) thawed-model))
+
+(ml/define-model! :standford-nlp/cfr
```
Comment (behrica) [inline src/dk/simongray/datalinguist/ml/crf.clj:107]: this registers the model with `tech.ml` on requiring this namespace

after this, the model can be used as any other model in `tech.ml`
See comments below

File: src/dk/simongray/datalinguist/ml/crf.clj [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  train
+  predict
+  {})
+
+
+(comment
```
Comment (behrica) [inline src/dk/simongray/datalinguist/ml/crf.clj:113]: this standrt `tech.ml` code to train / predict.

Same for any model.

File: src/dk/simongray/datalinguist/ml/crf.clj [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  (def train-ds
+    (->
+     (ds/->dataset "./test/data/jane-austen-emma-ch1.tsv" {:header-row? false :key-fn keyword})
+     (ds-mod/set-inference-target :column-1)))
+
+  (def model
+    (ml/train train-ds {:model-type :standford-nlp/cfr}))
```
Comment (behrica) [inline src/dk/simongray/datalinguist/ml/crf.clj:120]: choose the above registered model for training.
Could get any number of parameters 

`train` returns the trained model in a map with some standard keys.
(which needs to be passed into `predict` for predictions)

File: deps.edn [hunk 6]
```diff
@@ -2,7 +2,13 @@
...
+           techascent/tech.ml.dataset {:mvn/version "5.02"}
+           ;; I need this PR https://github.com/techascent/tech.ml.dataset/pull/216 as well, not released yet
```
Comment (behrica) [inline deps.edn:8]: but code works anyway

**Comment by simongray** [comment]
Thank you for making a PR.

I do have some comments:

* Stuff like`_` bindings inside a `let` seem a little bit sloppy, looks like temporary code to me. Perhaps you could use `doto` where applicable or refactor it a bit when `doto` doesn't cut it?
* I will also need some docstrings for the public fns at the very least. You've already made some comments about various parts of the code in this PR, so perhaps some of that could be refashioned into docstrings? I use the Markdown-like format where `name-of-param` is enclosed in backticks.
* Code formatting should try to follow the rest of the codebase. I try to use the [Clojure style guide](https://github.com/bbatsov/clojure-style-guide) with minor variations. Basically, I just use Cursive's autoformatting feature, with vertical alignment enabled. If you take a look at my code it shouldn't be too hard to see how I format stuff, since I'm pretty consistent about it.
* The Rich Comment block is fine, but since I really have no experience with the tech.ml stack it would be nice if that testing code could be formalised as an actual runnable test, so that it is made clear what the right result is supposed to be. I know that I haven't really made any tests myself, but this project - so far - consists entirely of my own code so I know what it's supposed to be doing.
* I don't think the clojurewerkz dependency is strictly necessary. If you need code for converting a map into props, I have [made some already](https://github.com/simongray/datalinguist/blob/master/src/dk/simongray/datalinguist.clj#L328-L358) that could be moved to a util ns of some sort. But perhaps the clojurewerkz stuff works better than mine, in which case I should probably delete my own code and just use that instead, but I'm not sure it necessarily does. Will have to take a look at it and report back to you.

I can perhaps help you make some of the changes (props conversion and any final touches on formatting), but you will need to work a little bit more on it yourself first, before I am ready to accept it. Hope that is okay with you. If that's too bothersome, just put it in your own repo instead :)

**Comment by behrica** [comment]
Thanks for review and suggestions.

I addressed all of them, except the Code formatting, as I don't use Cursive, but emacs.

I made the `properties` function in datalinguist.clj public for the moment and it does work.
Maybe you want to move this to a different ns. It works for me.

**Comment by simongray** [comment]
I took me a while to check out the changes. I just got a kid (he'll be 2 months tomorrow) so a lot of my sparetime is spent on him at the moment.

I made a lot of code style changes to make the code better fit in with the rest of the repo. The only major functional change is an introduction of a `define-model!` function, since I noticed that your code didn't provide a way to redefine the behaviour of the model using the options map (... without reaching into the source code).

I hope you agree with the changes. If you're ok with it, I'm ready to merge.

**Review by simongray — APPROVED** [review]

File: src/dk/simongray/datalinguist/ml/crf.clj [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  "Predict on new data using the CRF classifier.
+
+  The data need to be given a texts in the feature column of the dataset.
+
+  This method gets called indirectly by `tech.ml`."
+  [feature-dataset thawed-model model]
+  (let [word-col (first (ds/column-names feature-dataset))]
+    (predict-crf feature-dataset word-col thawed-model)))
+
+(defn define-model!
```
Comment (behrica) [inline src/dk/simongray/datalinguist/ml/crf.clj:102]: I think this is not in the spirit of the `tech.ml` architecture.
The "model registration` should happen ones.

But indeed we need to be able to change the configuration, but at runtime by calling `train` and pass options.
This happens in line 89, the :crf-options passed in will overwrite the default options.

This need to be documented, indeed.

'Tech.ml' has a way to document all options a model can take as a "map".
This is shown here:
https://github.com/scicloj/scicloj.ml.smile/blob/49a753c8ce6ae991b591a563f6a723c660ddb776/src/scicloj/ml/smile/sparse_logreg.clj#L48

I was planning to do this as well.

This metadata can then be used to generate documentation:
https://scicloj.github.io/scicloj.ml/userguide-models.html

**Comment by behrica** [comment]
It is as well foreseen to add specific metadata, which would point to:
https://nlp.stanford.edu/nlp/javadoc/javanlp/edu/stanford/nlp/ie/crf/CRFClassifier.html
and to a "user guide", if existing.

I have done this only for one model so far:

:smile.classification/gradient-tree-boost

https://github.com/scicloj/scicloj.ml.smile/blob/49a753c8ce6ae991b591a563f6a723c660ddb776/src/scicloj/ml/smile/classification.clj#L150

I would therefore propose to do the same here.
The doc-string of the "train" method can maybe contain "pointers" to this as well

File: src/dk/simongray/datalinguist/ml/crf.clj (hunk 7 above)
Comment (simongray) [reply src/dk/simongray/datalinguist/ml/crf.clj:102]: I see. I guess I should revert this to just {} at the moment, remove the `define-model!` fn and leave in a TODO for later?

**Comment by simongray** [comment]
> It is as well foreseen to add specific metadata, which would point to:
> https://nlp.stanford.edu/nlp/javadoc/javanlp/edu/stanford/nlp/ie/crf/CRFClassifier.html
> and to a "user guide", if existing.
> 
> I have done this only for one model so far:
> 
> :smile.classification/gradient-tree-boost
> 
> https://github.com/scicloj/scicloj.ml.smile/blob/49a753c8ce6ae991b591a563f6a723c660ddb776/src/scicloj/ml/smile/classification.clj#L150
> 
> I would therefore propose to do the same here.
> The doc-string of the "train" method can maybe contain "pointers" to this as well

Sure, this all sounds good to me. Feel free to do it in a follow-up PR once this has been merged.

File: src/dk/simongray/datalinguist/ml/crf.clj (hunk 7 above)
Comment (behrica) [reply src/dk/simongray/datalinguist/ml/crf.clj:102]: yes, just revert it.

I can add all parameters know as well, before you merge.

Comment (behrica) [reply src/dk/simongray/datalinguist/ml/crf.clj:102]: yes

**Comment by behrica** [comment]
Seems all good to me.
Please revert the "define.model!" changes and leave the todo for adding the options.
I will do this soon.

**Pull request merged**